python/ray/autoscaler/_private/kubernetes_operator/node_provider.py

```diff
@@ -0,0 +1,297 @@
+import logging
+from uuid import uuid4
+from kubernetes.client.rest import ApiException
+
+from ray.autoscaler._private.command_runner import KubernetesCommandRunner
+from ray.autoscaler._private.kubernetes_operator import core_api, log_prefix, \
+    extensions_beta_api, custom_object_api
+from ray.autoscaler._private.kubernetes_operator.config import bootstrap_kubernetes
+from ray.autoscaler.node_provider import NodeProvider
+from ray.autoscaler.tags import TAG_RAY_CLUSTER_NAME, TAG_RAY_NODE_KIND
+from ray.autoscaler._private.kubernetes_operator.ray_cluster_util import CRD_RAY_GROUP, \
+    CRD_RAY_VERSION, CRD_RAY_PLURAL, CRD_RAY_KIND, DASH
+
+logger = logging.getLogger(__name__)
+
+
+def to_label_selector(tags):
+    label_selector = ""
+    for k, v in tags.items():
+        if label_selector != "":
+            label_selector += ","
+        label_selector += "{}={}".format(k, v)
+    return label_selector
+
+
+class KubernetesOperatorNodeProvider(NodeProvider):
+    def __init__(self, provider_config, cluster_name):
+        NodeProvider.__init__(self, provider_config, cluster_name)
+        self.cluster_name = cluster_name
+        self.namespace = provider_config["namespace"]
+
+    def non_terminated_nodes(self, tag_filters):
+        # Match pods that are in the 'Pending' or 'Running' phase.
+        # Unfortunately there is no OR operator in field selectors, so we
+        # have to match on NOT any of the other phases.
+        field_selector = ",".join([
+            "status.phase!=Failed",
+            "status.phase!=Unknown",
+            "status.phase!=Succeeded",
+            "status.phase!=Terminating",
+        ])
+
+        tag_filters[TAG_RAY_CLUSTER_NAME] = self.cluster_name
+        label_selector = to_label_selector(tag_filters)
+        pod_list = core_api().list_namespaced_pod(
+            self.namespace,
+            field_selector=field_selector,
+            label_selector=label_selector)
+
+        return [pod.metadata.name for pod in pod_list.items]
+
+    def is_running(self, node_id):
+        pod = core_api().read_namespaced_pod(node_id, self.namespace)
+        return pod.status.phase == "Running"
+
+    def is_terminated(self, node_id):
+        pod = core_api().read_namespaced_pod(node_id, self.namespace)
+        return pod.status.phase not in ["Running", "Pending"]
```

Comment:
Is `Init` also one of states of non_terminated?

Reply:
- There are five possible phase values:  Pending, Running, Succeeded, Failed, Unknown.
- Pending: The pod has been accepted by the Kubernetes system, but one or more of the container images has not been created. This includes time before being scheduled as well as time spent downloading images over the network, which could take a while.